src/geometry/algebra/lie_group.lean

```diff
@@ -0,0 +1,275 @@
+the model space is `model_prod E E'` and the model vector space is `E × E'`, which are not the same,
+so the definition does not apply. Hence the definition should be more general, allowing
+`I : model_with_corners 𝕜 E H`.
+-/
+
+noncomputable theory
+
+section lie_group
+
+set_option default_priority 100
+
+/-- A Lie (additive) group is a group and a smooth manifold at the same time in which
+the addition and negation operations are smooth. -/
+class lie_add_group {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+  {H : Type*} [topological_space H]
+  {E : Type*} [normed_group E] [normed_space 𝕜 E] (I : model_with_corners 𝕜 E H)
+  (G : Type*) [add_group G] [topological_space G] [topological_add_group G] [charted_space H G]
+  extends smooth_manifold_with_corners I G : Prop :=
+(smooth_add : smooth (I.prod I) I (λ p : G×G, p.1 + p.2))
+(smooth_neg : smooth I I (λ a:G, -a))
+
+/-- A Lie group is a group and a smooth manifold at the same time in which
+the multiplication and inverse operations are smooth. -/
+@[to_additive]
+class lie_group {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+  {H : Type*} [topological_space H]
+  {E : Type*} [normed_group E] [normed_space 𝕜 E] (I : model_with_corners 𝕜 E H)
+  (G : Type*) [group G] [topological_space G] [topological_group G] [charted_space H G]
+  extends smooth_manifold_with_corners I G : Prop :=
+(smooth_mul : smooth (I.prod I) I (λ p : G×G, p.1 * p.2))
+(smooth_inv : smooth I I (λ a:G, a⁻¹))
+
+section lie_group
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{H : Type*} [topological_space H]
+{E : Type*} [normed_group E] [normed_space 𝕜 E] {I : model_with_corners 𝕜 E H}
+{F : Type*} [normed_group F] [normed_space 𝕜 F] {J : model_with_corners 𝕜 F F}
+{G : Type*} [topological_space G] [charted_space H G] [group G]
+[topological_group G] [lie_group I G]
+{E' : Type*} [normed_group E'] [normed_space 𝕜 E']
+{H' : Type*} [topological_space H'] {I' : model_with_corners 𝕜 E' H'}
+{M : Type*} [topological_space M] [charted_space H' M] [smooth_manifold_with_corners I' M]
+{E'' : Type*} [normed_group E''] [normed_space 𝕜 E'']
+{H'' : Type*} [topological_space H''] {I'' : model_with_corners 𝕜 E'' H''}
+{M' : Type*} [topological_space M'] [charted_space H'' M'] [smooth_manifold_with_corners I'' M']
+
+@[to_additive]
+lemma smooth_mul : smooth (I.prod I) I (λ p : G×G, p.1 * p.2) :=
+lie_group.smooth_mul
+
+@[to_additive]
+lemma smooth.mul {f : M → G} {g : M → G} (hf : smooth I' I f) (hg : smooth I' I g) :
+  smooth I' I (f * g) :=
+smooth_mul.comp (hf.prod_mk hg)
+
+namespace lie_group
+
+/-- `L g` denotes left multiplication by `g` -/
+abbreviation L : G → G → G := left_mul g
+
+/-- `R g` denotes right multiplication by `g` -/
+abbreviation R : G → G → G := right_mul g
+
+attribute [nolint unused_arguments] lie_add_group.R_add
+
+end lie_group
+
+@[to_additive]
+lemma smooth_mul_left {a : G} : smooth I I (λ b : G, a * b) :=
```

Comment:
Since the operations are called `left_mul` and `right_mul`, the lemma names should probably match.